sources/agent/greynoise/greynoise.go

```diff
@@ -0,0 +1,187 @@
+				return
+			}
+			if apiResponse == nil || len(apiResponse.Data) == 0 {
+				return
+			}
+
+			for _, item := range apiResponse.Data {
+				host := firstNonEmpty(
+					item.InternetScannerIntelligence.Metadata.Domain,
+					item.InternetScannerIntelligence.Metadata.RDNS,
+				)
+
+				r := sources.Result{
+					Source: agent.Name(),
+					IP:     item.IP,
+					Host:   host,
+				}
+				if raw, err := json.Marshal(item); err == nil {
+					r.Raw = raw
+				}
+				results <- r
+
+				total++
+				if query.Limit > 0 && total >= query.Limit {
+					done = true
+					break
+				}
+			}
+
+			done = done || apiResponse.RequestMetadata.Complete
+			scrollToken = apiResponse.RequestMetadata.Scroll
+			if strings.TrimSpace(scrollToken) == "" {
+				done = true
+			}
+
+			if query.Limit > 0 && !done {
+				remain := query.Limit - total
+				if remain < pageSize {
+					pageSize = remain
+				}
+			}
+		}
+	}()
+
+	return results, nil
+}
+
+func (agent *Agent) query(session *sources.Session, request *Request) (*Response, error) {
+	params := url.Values{}
+	params.Set("query", request.Query)
+
+	if request.Size > 0 {
+		if request.Size > 10000 {
+			request.Size = 10000
+		}
+		params.Set("size", strconv.Itoa(request.Size))
+	}
+	if request.Scroll != "" {
+		params.Set("scroll", request.Scroll)
+	}
+	path := "/v3/gnql"
+	if request.ExcludeRaw {
+		path = "/v3/gnql/metadata"
+	}
+
+	baseURL, _ := url.Parse(URL)
+	baseURL.Path = path
+	fullURL := baseURL.String()
+	if enc := params.Encode(); enc != "" {
+		fullURL = fullURL + "?" + enc
+	}
+
+	req, err := sources.NewHTTPRequest(http.MethodGet, fullURL, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Set("Accept", "application/json")
+	req.Header.Set("key", session.Keys.GreyNoiseKey)
+
+	resp, err := session.Do(req, agent.Name())
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode < 200 || resp.StatusCode > 299 {
+		b, _ := io.ReadAll(resp.Body)
+		msg := strings.TrimSpace(string(b))
+
+		if resp.StatusCode == http.StatusUnauthorized || resp.StatusCode == http.StatusForbidden || resp.StatusCode == http.StatusNotFound {
+			return nil, fmt.Errorf(
+				"GreyNoise GNQL request failed: status=%d. Your API key may not include GNQL access (Enterprise key required). body=%s",
+				resp.StatusCode, msg,
+			)
+		}
+
+		return nil, fmt.Errorf("greynoise GNQL request failed: status=%d body=%s", resp.StatusCode, msg)
+	}
+
+	var apiResponse Response
+	if err := json.NewDecoder(resp.Body).Decode(&apiResponse); err != nil {
+		fmt.Fprintf(os.Stderr, "DEBUG: GreyNoise decode error status=%d: %v\n", resp.StatusCode, err)
+		return nil, err
+	}
+
+	fmt.Fprintf(os.Stderr,
+		"DEBUG: GNQL count=%d complete=%v scroll=%s data=%d msg=%s\n",
+		apiResponse.RequestMetadata.Count,
+		apiResponse.RequestMetadata.Complete,
+		short(apiResponse.RequestMetadata.Scroll, 12),
+		len(apiResponse.Data),
+		short(apiResponse.RequestMetadata.Message, 120),
+	)
```

Comment:
this should only be displayed if debug's enabled.

also uses `gologger.Debug()` instead.